README.md
# MantisBT OnCallAlerts Plugin

* [English](#english)
* [Français](#french)

## English
<a name="english"></a>

__Overview__
On-call management plugin and SMS alerts sending.

__Features__

* On-call management and technician assignments.
* Issues monitoring and SMS sending to the technicians.
* Manage a monitoring plan.
* SMS alert sending before each assignment (need a [cron](https://en.wikipedia.org/wiki/Cron) task ).

__Requirements__

* [MantisCore => 1.2.0](https://sourceforge.net/projects/mantisbt/files/mantis-stable/1.2.0/)
* [Query => 1.6](https://github.com/mantisbt-plugins/jquery)
* [jQueryUI  => 1.11.4](https://github.com/mantisbt-plugins/jQuery-UI)

In order to send SMS alerts, you will need a SMS Server.

__Installation__

1. Download and unzip the plugin files to your computer.
2. Upload the plugin directory and the files it contains under: *yourMantisRoot/plugins*.
3. In MantisBT, go to *Manage > Manage Plugins*. You will see a list of installed and not installed plugins.
4. Click the *Install* link to install a plugin.

[More information](https://www.mantisbt.org/wiki/doku.php/mantisbt:mantis_plugins)

__How to use__

1. In MantisBT, go to *Manage > Manage Plugins*.
2. Click on the plugin named *On Call Alerts*.
3. Fill in the technical configuration form (SMS messages, SMS server configuration, etc.).
4. Click on *Configuration On Call Alerts* to configure and use your plugin.

__Contributing__

1. Fork the project
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

##  Français  
<a name="french"></a>

___Description___
Module de gestion d'astreinte et envoi d'alertes SMS.

__Fonctionnalités__

* Gestion d'astreintes et assignation de techniciens.
* Surveillance des bogues et envoi de SMS selon les astreintes du jour.
* Gestion d'un planning de surveillance.
* Rappel par SMS avant chaque astreinte (nécessite une tâche [cron](https://fr.wikipedia.org/wiki/Cron)).

__Prérequis__

* [MantisCore => 1.2.0](https://sourceforge.net/projects/mantisbt/files/mantis-stable/1.2.0/)
* [Query => 1.6](https://github.com/mantisbt-plugins/jquery)
* [jQueryUI  => 1.11.4](https://github.com/mantisbt-plugins/jQuery-UI)

Note : un serveur SMS est également nécessaire pour la gestion des alertes par SMS.
Aesis Conseil peut vous fournir un serveur SMS si vous n'en avez pas.
Merci de nous contacter : https://www.aesis-conseil.com/contact-aesis/

__Installation__

1. Téléchargez et décompressez le fichier du plugin sur votre ordinateur
2. Téléversez le répertoire du plugin et les fichiers qu'il contient sous : *votreRepertoireMantis/plugins*
3. Dans MantisBT, accédez à la page *Gérer > Gérer les plugins*. Vous verrez une liste des plugins installés et non installés.
4. Cliquez sur le lien *Installer* pour installer un plugin.

[Plus d'informations sur les plugins dans Mantis](https://www.mantisbt.org/wiki/doku.php/mantisbt:mantis_plugins)

__Utilisation__

1. Dans MantisBT, accédez à la page *Gérer > Gérer les plugins*.
2.  Cliquez sur le nom du plugin *On Call Alerts*
3. Remplissez le formulaire de configuration technique (messages des SMS, configuration du serveur SMS etc.)
4.  Cliquez sur *Configuration On Call Alerts* et configurez vos techniciens, astreintes et planning.

__Contribuer__

1. Créez un fork du projet
2. Crééz votre branche (`git checkout -b my-new-feature`)
3. Commit vos changements (`git commit -am 'Add some feature'`)
4. Push dans la branche (`git push origin my-new-feature`)
5. Créez une nouvelle Pull Request



changelog.md
**ChangeLog**

** Vers 1.2 **

* MAj des fichier de langues et corrections. 


** Vers 1.1 **

* Correction d'un bug sur le tri des dates. Celui-ci devrait maintenant être correct.
* Correction du bug sur l'url du site mobile lors de l'envoi d'un sms. Si une url mobile est configuré dans 
les paramêtre, l'url dans le sms devrait maintenant être la bonne.
* Correction d'un bug concernant l'ajout d'un numéro de téléphone qui permettait d'ajouter
un numéro incorrect

* Ajout de l'édition du numero de téléphone d'un technicien afin de ne pas perdre l'historique en cas 
de changement de numéro de téléphone. 
* Ajout d'une page PHP CLI qui envoie un sms de rappel au technicien en astreinte à J+1. Afin 
d'utiliser cette fonctionnalitée, faire appel au fichier : files/phpcli_sendText_recall.php toute les 24h. 

OnCallAlerts.php
<?php

/* 
 * Plugin OnCallAlerts  pour Mantis BugTracker 
 *
 */

class OnCallAlertsPlugin extends MantisPlugin {

    function register() {
        $this->name         = 'On Call Alerts';
        $this->description  = lang_get('plugin_OnCallAlerts_description');
        $this->version      = "1.2";
        $this->requires     = array(
            "MantisCore" => "1.2.0",
            "jQuery"     => "1.6",
            "jQueryUI"   => "1.11.4",
        );
        $this->page         = 'config';
        $this->author       = "DEV : Thomas SIMON | Graphisme : Laure MARTINEZ | Translate : Rémi Nedelec";
        $this->contact      = "tsimon@aesis-conseil.com";
        $this->url          = "http://www.aesis-conseil.com";
    }

    function config() { 
        return array(
            'nb_day_log'            => 20,
            'msg_text'              => plugin_lang_get('default_text'),
            'msg_text_rappel'       => plugin_lang_get('default_text_rappel'),
            'msg_text_test'         => plugin_lang_get('default_text_test'),
            'url_mantis_text'       => 'https://support.aesis-conseil.com',
            'url_server_text'       => '',
            'login_server_text'     => '',
            'pwd_server_text'       => '',
            'publickey_server_text' => '',
            'data_server_text'      => '',
            'select_project'        => array('0'),
            'newBug'                => '',
            'select_newbug'         => array(),
            'updateBug'             => '',
            'select_updateBug'      => array(),
            'addNote'               => '',
            'select_addNote'        => array(), 
        );
    }

    function schema() {
        return array(
            array("CreateTableSQL", array(plugin_table("oncall"), "
				id              I           NOTNULL UNSIGNED AUTOINCREMENT PRIMARY,
				technician_id   I           NOTNULL UNSIGNED,
				start_date      DateTime    NOTNULL,
				end_date        DateTime    NOTNULL,
                                note            XL               
				",
                    array("mysql" => "DEFAULT CHARSET=utf8"))),
            array("CreateTableSQL", array(plugin_table("planing"), "
				id              I       NOTNULL UNSIGNED AUTOINCREMENT PRIMARY,
				technician_id   I	NOTNULL UNSIGNED,
                                monday          BOOLEAN ,
                                tuesday         BOOLEAN ,
                                wednesday       BOOLEAN ,
                                thursday        BOOLEAN ,
                                friday          BOOLEAN ,
                                saturday        BOOLEAN ,
                                sunday          BOOLEAN ,
				start_time      Time    NOTNULL,
				end_time        Time	NOTNULL,
                                note            XL               
				",
                    array("mysql" => "DEFAULT CHARSET=utf8"))),
            array("CreateTableSQL", array(plugin_table("technician"), "
				id      I       NOTNULL UNSIGNED AUTOINCREMENT PRIMARY,
				name    C(50)	NOTNULL,
                                phone   C(20)   NOTNULL  
				",
                    array("mysql" => "DEFAULT CHARSET=utf8"))),
        );
    }

    function api() {
		require_once("api/Oncall.class.php");
		require_once("api/Planing.class.php");
		require_once("api/Technician.class.php");
		require_once("api/SendText.class.php");
		require_once("api/RSA/Crypt/RSA.php");
		require_once("api/Logger.class.php");                
    }
        
    function hooks() {
        $hooks = array(
	    'EVENT_CORE_READY'      => 'api',
            'EVENT_MENU_MANAGE' => 'print_menu_OnCallAlertsManagement',
            'EVENT_REPORT_BUG'  => 'send_sms_reported',
            'EVENT_UPDATE_BUG'  => 'send_sms_updated',
            'EVENT_BUGNOTE_ADD' => 'send_sms_addNote'
        );
        return $hooks;
    }

    function print_menu_OnCallAlertsManagement() {

        $page = plugin_page('management');
        $lang = plugin_lang_get('management_link');

        return "<a href='$page'>$lang</a>";
    }

    function send_sms_reported($event, $data, $bugID) {

        if(plugin_config_get('newBug'))
        { 
            if(array_search($data->severity,plugin_config_get('select_newbug')) !== false &&
               ( array_search($data->project_id,plugin_config_get('select_project')) !== false || array_search('0',plugin_config_get('select_project')) !== false )
              )
            {                
                $projectName = project_get_name($data->project_id);
                $reporterData = user_get_row($data->reporter_id);
                $categoryName = category_full_name($data->category_id);

                $planing = new OnCallAlertsPlaning();
                $res = $planing->getInPlaning();
                if($res)
                {
                    foreach($res as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
                
                $onCall = new OnCallAlertsOncall();
                $resOnCall = $onCall->getInOnCall();
               
                if($resOnCall)
                {
                    foreach($resOnCall as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
            }
        }
    }

    function send_sms_updated($event, $data, $bugID) {
        if(plugin_config_get('updateBug'))
        { 
            if(array_search($data->severity,plugin_config_get('select_updateBug')) !== false &&
               ( array_search($data->project_id,plugin_config_get('select_project')) !== false || array_search('0',plugin_config_get('select_project')) !== false )
              )
            {                
                $projectName = project_get_name($data->project_id);
                $reporterData = user_get_row($data->reporter_id);
                $categoryName = category_full_name($data->category_id);

                $planing = new OnCallAlertsPlaning();
                $res = $planing->getInPlaning();
                if($res)
                {
                    foreach($res as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
                
                $onCall = new OnCallAlertsOncall();
                $resOnCall = $onCall->getInOnCall();
               
                if($resOnCall)
                {
                    foreach($resOnCall as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
            }
        }
    }

    function send_sms_addNote($event, $bugID, $bugNoteID) {
        
        $data = bug_get($bugID, true);
        if(plugin_config_get('addNote'))
        { 
            if(array_search($data->severity,plugin_config_get('select_addNote')) !== false &&
               ( array_search($data->project_id,plugin_config_get('select_project')) !== false || array_search('0',plugin_config_get('select_project')) !== false )
              )
            {                
                $projectName = project_get_name($data->project_id);
                $reporterData = user_get_row($data->reporter_id);
                $categoryName = category_full_name($data->category_id);

                $planing = new OnCallAlertsPlaning();
                $res = $planing->getInPlaning();
                if($res)
                {
                    foreach($res as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
                
                $onCall = new OnCallAlertsOncall();
                $resOnCall = $onCall->getInOnCall();
               
                if($resOnCall)
                {
                    foreach($resOnCall as $r)
                    {
                        $date = new \DateTime();                        
                        $this->_sendText($r->phone, $categoryName, $bugID, $projectName, $reporterData['username'],$date->format('d-m-Y H:i'));
                        Logger::trace("Text send to {$r->phone} for new report by {$reporterData['username']} for $projectName");
                    }
                } 
            }
        }
    }

    private function _sendText($phone,$typeAlerte,$idBug,$projectName,$rapporteur, $date)
    {
        $text = new OnCallAlertsSendText();
        
        $aSearch = array('#CATEGORY','#IDBUG','#PROJECTNAME','#RAPPORTEUR','#DATE','#URL');
        
        $sUrl= plugin_config_get('url_mantis_text');
        
        if(substr($sUrl,-1,1) != '/') $sUrl = $sUrl.'/';
        
        $findme    = '/m/';
        
        if( stripos($sUrl, $findme) ) $rUrl = $sUrl.'issue_page.php?id='.$idBug;
        else $rUrl = $sUrl.'view.php?id='.$idBug;
        
        $aReplace = array($typeAlerte,$idBug,$projectName,$rapporteur, $date,$rUrl);
        
        $msgText = substr(str_replace($aSearch,$aReplace,plugin_config_get('msg_text')),0,159);
       
        if($text->sendText($msgText, $phone))
        {
            return true;
        }
        else return false;
    }
}